## Question & Answering using Transformers

### Installation

In [1]:
!pip install transformers

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 7.4/7.4 MB 23.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 268.8/268.8 kB 29.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 7.8/7.8 MB 56.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.3/1.3 MB 62.7 MB/s eta 0:00:00


In [3]:
!pip install datasets

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 519.3/519.3 kB 5.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 115.3/115.3 kB 7.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 194.1/194.1 kB 9.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 134.8/134.8 kB 12.2 MB/s eta 0:00:00


## Approach-1: Using custom context

### Import required modules

In [15]:
from transformers import pipeline
import warnings
warnings.filterwarnings("ignore")

In [16]:
# Load the Q&A pipeline
qa_pipeline = pipeline("question-answering") #this model uses Distil-BERT Cased model

No model was supplied, defaulted to distilbert-base-cased-distilled-squad and revision 626af31 (https://huggingface.co/distilbert-base-cased-distilled-squad).
Using a pipeline without specifying a model name and revision in production is not recommended.


No CUDA runtime is found, using CUDA_HOME='/usr/local/cuda'


> #### In the Module - 5 we will learn about BERT in depth

In [17]:
# Example context
context = """
AI Planet is a global AI community with headquarters in Belgium and India.
It started with the vision to make AI education accessible to everyone and build AI for good to solve key challenges of humanity.
As part of our community initiatives, we provide free AI and data science courses by industry experts from large tech companies or
startups worldwide. Over 300K+ learners from 150+ countries have benefited since our inception in 2020.
"""

In [18]:
# Ask questions based out of the context
question1 = "How many learners have AI Planet impacted?"
question2 = "Where is AI Planet located?"
question3 = "What is the vision of AI Planet?"

In [19]:
#Use Question and Answering by default
def answer_question(context, question):

    answer = qa_pipeline(context=context, question=question)
    return answer["answer"]

In [20]:
# Get the answer
answer1 = answer_question(context, question1)
answer2 = answer_question(context,question2)
answer3 = answer_question(context,question3)

In [21]:
print(f"Question:{question1}")
print(f"Answer:{answer1}")

Question:How many learners have AI Planet impacted?
Answer:Over 300K+


In [22]:
print(f"Question:{question2}")
print(f"Answer:{answer2}")

Question:Where is AI Planet located?
Answer:Belgium and India


In [23]:
print(f"Question:{question3}")
print(f"Answer:{answer3}")


Question:What is the vision of AI Planet?
Answer:to make AI education accessible to everyone


## Approach-2: Using datasets from HuggingFace

In [4]:
import datasets

### Load the dataset using `load_dataset`

We will use the dataset names **squad**. This dataset contain, Context, Question and Answer features

In [5]:
dataset = datasets.load_dataset("squad")

Extracting data files:   0%|          | 0/2 [00:00<?, ?it/s]

Generating train split:   0%|          | 0/87599 [00:00<?, ? examples/s]

Generating validation split:   0%|          | 0/10570 [00:00<?, ? examples/s]

### Visualize the dataset

It is split into two part:
- Training dataset
- Validation dataset

In [6]:
dataset

DatasetDict({
    train: Dataset({
        features: ['id', 'title', 'context', 'question', 'answers'],
        num_rows: 87599
    })
    validation: Dataset({
        features: ['id', 'title', 'context', 'question', 'answers'],
        num_rows: 10570
    })
})

The dataset contains: 5 columns- id, title, context, question and answers. The training data include 87,599 rows whereas validation dataset includes 10,570 rows. By rows it means total number of context, questions and answers

In [24]:
dataset['train']['title'][:10]

['University_of_Notre_Dame',
 'University_of_Notre_Dame',
 'University_of_Notre_Dame',
 'University_of_Notre_Dame',
 'University_of_Notre_Dame',
 'University_of_Notre_Dame',
 'University_of_Notre_Dame',
 'University_of_Notre_Dame',
 'University_of_Notre_Dame',
 'University_of_Notre_Dame']

In [9]:
dataset['train']['question'][:10]

['To whom did the Virgin Mary allegedly appear in 1858 in Lourdes France?',
 'What is in front of the Notre Dame Main Building?',
 'The Basilica of the Sacred heart at Notre Dame is beside to which structure?',
 'What is the Grotto at Notre Dame?',
 'What sits on top of the Main Building at Notre Dame?',
 'When did the Scholastic Magazine of Notre dame begin publishing?',
 "How often is Notre Dame's the Juggler published?",
 'What is the daily student paper at Notre Dame called?',
 'How many student news papers are found at Notre Dame?',
 'In what year did the student paper Common Sense begin publication at Notre Dame?']

In [10]:
dataset['train']['answers'][:10]

[{'text': ['Saint Bernadette Soubirous'], 'answer_start': [515]},
 {'text': ['a copper statue of Christ'], 'answer_start': [188]},
 {'text': ['the Main Building'], 'answer_start': [279]},
 {'text': ['a Marian place of prayer and reflection'], 'answer_start': [381]},
 {'text': ['a golden statue of the Virgin Mary'], 'answer_start': [92]},
 {'text': ['September 1876'], 'answer_start': [248]},
 {'text': ['twice'], 'answer_start': [441]},
 {'text': ['The Observer'], 'answer_start': [598]},
 {'text': ['three'], 'answer_start': [126]},
 {'text': ['1987'], 'answer_start': [908]}]

In [11]:
dataset['train']['context'][:10]

['Architecturally, the school has a Catholic character. Atop the Main Building\'s gold dome is a golden statue of the Virgin Mary. Immediately in front of the Main Building and facing it, is a copper statue of Christ with arms upraised with the legend "Venite Ad Me Omnes". Next to the Main Building is the Basilica of the Sacred Heart. Immediately behind the basilica is the Grotto, a Marian place of prayer and reflection. It is a replica of the grotto at Lourdes, France where the Virgin Mary reputedly appeared to Saint Bernadette Soubirous in 1858. At the end of the main drive (and in a direct line that connects through 3 statues and the Gold Dome), is a simple, modern stone statue of Mary.',
 'Architecturally, the school has a Catholic character. Atop the Main Building\'s gold dome is a golden statue of the Virgin Mary. Immediately in front of the Main Building and facing it, is a copper statue of Christ with arms upraised with the legend "Venite Ad Me Omnes". Next to the Main Building

In [43]:
context = dataset["train"][0]["context"]
question = "Where is the modern stone statue of Mary"

In [44]:
answer = qa_pipeline(question=question, context=context)

In [45]:
print(context)

Architecturally, the school has a Catholic character. Atop the Main Building's gold dome is a golden statue of the Virgin Mary. Immediately in front of the Main Building and facing it, is a copper statue of Christ with arms upraised with the legend "Venite Ad Me Omnes". Next to the Main Building is the Basilica of the Sacred Heart. Immediately behind the basilica is the Grotto, a Marian place of prayer and reflection. It is a replica of the grotto at Lourdes, France where the Virgin Mary reputedly appeared to Saint Bernadette Soubirous in 1858. At the end of the main drive (and in a direct line that connects through 3 statues and the Gold Dome), is a simple, modern stone statue of Mary.


In [46]:
print(f"Question:{question}")

Question:Where is the modern stone statue of Mary


In [47]:
print(f"Answer:{answer}")

Answer:{'score': 0.7316837310791016, 'start': 551, 'end': 579, 'answer': 'At the end of the main drive'}
